src/hotspot/share/oops/trainingData.cpp

```diff
@@ -624,6 +627,18 @@ void CompileTrainingData::verify() {
     }
     guarantee(ktd->_comp_deps.contains(this), "");
   }
+
+  if (verify_dep_counter) {
+    int init_deps_left1 = init_deps_left_acquire();
+    int init_deps_left2 = compute_init_deps_left();
+
+    bool invariant = (init_deps_left1 >= init_deps_left2);
```

Comment:
I assume this check takes concurrent class initialization into account and init notification events are processed on a dedicated thread. Can we strengthen the check by repeatedly performing it and ensuring the value converges? Also, maybe take event queue into account?

Reply:
It's very hard to do reliably given the way the vm shutdown currently works. There is no way to ensure that all the java threads are stopped, so checking the convergence is problematic. So, the best I can do right now is prove the `>=` property.

Reply:
I mean, I tired, but gave up on the convergence for now. Perhaps we'd make a stab at it another time.

Reply:
Ok, sounds good. Thanks for the clarifications.